crypto/merklearray/layer.go

```diff
@@ -0,0 +1,40 @@
+package merklearray
+
+import (
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// A layer of the Merkle tree consists of a dense array of hashes at that
+// level of the tree.  Hashes beyond the end of the array (e.g., if the
+// number of leaves is not an exact power of 2) are implicitly zero.
+type layer []crypto.Digest
+
+// A pair represents an internal node in the Merkle tree.
+type pair struct {
+	l crypto.Digest
+	r crypto.Digest
+}
+
+func (p *pair) ToBeHashed() (protocol.HashID, []byte) {
+	var buf [2 * crypto.DigestSize]byte
+	copy(buf[:crypto.DigestSize], p.l[:])
+	copy(buf[crypto.DigestSize:], p.r[:])
+	return protocol.MerkleArrayNode, buf[:]
+}
+
+// up takes a layer representing some level in the tree,
+// and returns the next-higher level in the tree,
+// represented as a layer.
+func (l layer) up() layer {
+	res := make(layer, (len(l)+1)/2)
```

Comment:
I am not sure how reasonable of a concern this is, but on a 32-bit platform then I think this could plausibly overflow if `l` has length `(2 << 31) - 1`

https://play.golang.org/p/W-PmprYmWEc

Reply:
Isn't that intended to contain the transactions within a block ? if so, I can't imagine that we'll reach more than 2^31 tx/block. or maybe I miss something ?

Reply:
Yeah, I was just reviewing this as an abstract primitive. If something at a higher level is limiting the length then it's nothing to worry about.

Reply:
Sure, might as well fix it!  Thanks!